rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,416 @@
+
+use crate::leased_lock::{Error, ErrorKind, LeaseObservation, Response, SetCondition, SetOptions};
+use crate::state_store::{self};
+use azure_iot_operations_mqtt::interface::ManagedClient;
+use azure_iot_operations_protocol::common::hybrid_logical_clock::HybridLogicalClock;
+
+/// Key Lease client struct.
+pub struct Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync + 'static,
+    C::PubReceiver: Send + Sync,
+{
+    state_store: Arc<state_store::Client<C>>,
+    key_name: Vec<u8>,
+    lease_holder_name: Vec<u8>,
+    current_fencing_token: Arc<Mutex<Option<HybridLogicalClock>>>,
+    auto_renewal_task: JoinHandle<()>,
+}
+
+/// Key Lease client implementation
+///
+/// Notes:
+/// Do not call any of the methods of this client after the `state_store` parameter is shutdown.
+/// Calling any of the methods in this implementation after the `state_store` is shutdown results in undefined behavior.
+impl<C> Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync,
+    C::PubReceiver: Send + Sync,
+{
+    /// Create a new Key Lease Client.
+    ///
+    /// Notes:
+    /// - `lease_holder_name` is expected to be the client ID used in the underlying MQTT connection settings.
+    /// - There must be one instance of `leased_lock::Client` per lease.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`KeyLengthZero`](ErrorKind::KeyLengthZero) if the `key_name` is empty
+    ///
+    /// [`struct@Error`] of kind [`LeaseHolderNameLengthZero`](ErrorKind::LeaseHolderNameLengthZero) if the `lease_holder_name` is empty
+    pub fn new(
+        state_store: Arc<state_store::Client<C>>,
+        key_name: Vec<u8>,
+        lease_holder_name: Vec<u8>,
+    ) -> Result<Self, Error> {
+        if key_name.is_empty() {
+            return Err(Error(ErrorKind::KeyLengthZero));
+        }
+
+        if lease_holder_name.is_empty() {
+            return Err(Error(ErrorKind::LeaseHolderNameLengthZero));
+        }
+
+        Ok(Self {
+            state_store,
+            key_name,
+            lease_holder_name,
+            current_fencing_token: Arc::new(Mutex::new(None)),
+            auto_renewal_task: tokio::task::spawn(async move {}), // Completes immediately, creating a finished JoinHandle.
+        })
+    }
+
+    /// Gets the latest fencing token related to the most recent lease.
+    ///
+    /// Returns either None or an actual Fencing Token (`HybridLogicalClock`).
+    /// None means that either a lease has not been acquired previously with this client, or
+    /// if a lease renewal has failed (if lease auto-renewal is used). The presence of a `HybridLogicalClock`
+    /// does not mean that it is the most recent (and thus valid) Fencing Token - in case
+    /// auto-renewal has not been used and the lease has already expired.
+    #[must_use]
+    pub async fn get_current_lease_fencing_token(&self) -> Option<HybridLogicalClock> {
+        self.current_fencing_token.clone().lock().await.clone()
```

Comment:
I think this first clone shouldn't be needed, it feels a little weird to be cloning the mutex

Reply:
It's cloning the Arc I think. But, yes, I agree, this seems very unnecessary

Reply:
This all has been reworked. Please review the new code.